### PR TITLE
Refactor/simplify scheduling extension

### PR DESCRIPTION
**Pull request by hkdsun**
Following the discussion in https://github.com/Shopify/resque-scheduler/pull/7 I started refactoring how schedules are stored. 

I don't think this should necessarily block us and we should still merge #7 if this takes a while to get to an acceptable state but it would be cool to get some feedback on this

@dylanahsmith @Sirupsen

### CODE REVIEW

**Comment by hkdsun** [comment]
The main problems that it solves are:
* We keep non-persistent schedules only in memory
* The startup race condition is fixed
* Schedules are prefixed with the master lock prefix

**Comment by hkdsun** [comment]
TL;DR dynamic scheduling is done best using a single source of truth

I think I just realized why the schedules were in Redis before :sob:, even though it does not really apply to us.

In dynamic scheduling (you can add/remove schedules using `Resque.set_schedule` or `Resque.remove_schedule`) we have a bunch of nodes running on a uniform schedule. If somebody wants to update the schedule dynamically from anywhere (say [through an HTTP request](https://github.com/Shopify/resque-scheduler/blob/startup-race-condition-2-develop/lib/resque/scheduler/server.rb#L27)) then the sensible way to make sure that the schedules stay uniform across all instances is to change one source of truth (`:schedules` on redis) and then making sure that any scheduler that is the master keeps itself up-to-date with that one source of truth, [perhaps in every polling loop](https://github.com/Shopify/resque-scheduler/blob/startup-race-condition-2-develop/lib/resque/scheduler.rb#L57) where [we call `reload_schedule!`](https://github.com/Shopify/resque-scheduler/blob/startup-race-condition-2-develop/lib/resque/scheduler.rb#L304).

In this PR I store the non-persistent schedule in memory for every node and that works as long as we don't give the user the power to _change_ the schedule while a scheduler is running (this is our use case) but as soon as we allow dynamically setting schedules this doesn't work anymore.

**Comment by hkdsun** [comment]
closing this for now, unless i come up with something else later on

File: lib/resque/scheduler.rb [hunk 1]
```diff
@@ -92,7 +92,7 @@ def load_schedule!
         Resque.schedule.each do |name, config|
           load_schedule_job(name, config)
         end
-        Resque.redis.del(:schedules_changed)
+        Resque.redis.del(Resque.keys[:schedules_changed])
```
Comment (dylanahsmith) [inline lib/resque/scheduler.rb]: Never use Resque.keys, it is O(N) for the entire keyspace.

Comment (hkdsun) [inline lib/resque/scheduler.rb]: I was referring to the [proxy method i defined](https://github.com/Shopify/resque-scheduler/pull/8/files#diff-74cb42b5aa9ca904352848778d8e0ce4R112). Did you mean `Resque.redis.keys`?

Comment (dylanahsmith) [inline lib/resque/scheduler.rb]: I was referring to https://github.com/resque/resque/blob/v1.26.0/lib/resque.rb#L496-L502

Comment (hkdsun) [inline lib/resque/scheduler.rb]: hmm that's shitty i'll need to change the method name

**Comment by dylanahsmith** [comment]
> In this PR I store the non-persistent schedule in memory for every node and that works as long as we don't give the user the power to change the schedule while a scheduler is running (this is our use case) but as soon as we allow dynamically setting schedules this doesn't work anymore.

Doesn't `all_schedules` being defined as `non_persistent_schedules.merge(persistent_schedules)` mean  that setting non-persistent schedules does work.  I think that would preserve the previous behaviour.

**Comment by hkdsun** [comment]
`non_persistent_schedules` is stored in an instance's memory so it does work for that single instance. 

Imagine you have an instance running with schedule `S1`, you mutate its schedule so that you have `S2`, and then 10 minutes later you failover to a passive backup. The passive backup will still be on schedule `S1`.

**Comment by dylanahsmith** [comment]
Persistent changes would apply to the other instance using the schedules_changed key and the dynamic flag which reloads the schedule if anything was added to the schedules_changed set.  Ideally we wouldn't even bother touching the schedules_changed when not using the dynamic configuration option, since I think that is its only purpose.

**Comment by hkdsun** [comment]
So, we're saying that non-persistent schedules are local to a single node and not its backup?

File: lib/resque/scheduler/scheduling_extensions.rb [hunk 2]
```diff
@@ -121,44 +80,71 @@ def non_persistent_schedules
...
-        redis.sadd(:schedules_changed, name)
+        non_persistent_schedules.delete(name)
+        redis.hdel(keys[:persistent_schedules], name)
+        redis.sadd(keys[:schedules_changed], name)
+
+        reload_schedule!
+      end
+
+      def keys
+        lock_prefix = ENV['RESQUE_SCHEDULER_MASTER_LOCK_PREFIX'] || ''
```
Comment (dylanahsmith) [inline lib/resque/scheduler/scheduling_extensions.rb]: Maybe this change should be handled in a separate PR.  I'm not really clear on why just the lock would be namespaced, rather than all the scheduler's data or all of the resque redis data.  Either way, it seems like this ENV var isn't appropriately named after a change like this, and should probably be read with the other ENV vars in Resque::Scheduler::Configuration

**Comment by dylanahsmith** [comment]
> So, we're saying that non-persistent schedules are local to a single node and not its backup?

That behaviour does sound more true to its name. The schedule is going to get lost whenever it is cleaned anyways.

**Comment by dylanahsmith** [comment]
I think this PR should be re-opened, it seems like the right approach.

File: resque-scheduler.gemspec [hunk 3]
```diff
@@ -28,6 +28,7 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'minitest'
   spec.add_development_dependency 'mocha'
   spec.add_development_dependency 'pry'
+  spec.add_development_dependency 'byebug'
```
Comment (dylanahsmith) [inline resque-scheduler.gemspec]: I think byebug is MRI specific, so this should probably be put in the Gemfile as `gem 'byebug', :platforms => :mri`

**Comment by hkdsun** [comment]
okay let's do this :)

**Comment by hkdsun** [comment]
cool. i removed the changes related to lock-based namespacing

**Comment by dylanahsmith** [comment]
:+1:

**Comment by hkdsun** [comment]
all shopify tests pass on this branch

File: lib/resque/scheduler/scheduling_extensions.rb [hunk 4]
```diff
@@ -121,44 +80,61 @@ def non_persistent_schedules
...
-      def prepare_schedule(schedule_hash)
-        prepared_hash = {}
+      # we store our non-persistent schedules in this hash
+      def non_persistent_schedules
+        @non_persistent_schedules ||= {}
+      end
+
+      # reads the persistent schedules from redis
+      def persistent_schedules
+        redis.hgetall(:persistent_schedules).tap do |h|
```
Comment (Sirupsen) [inline lib/resque/scheduler/scheduling_extensions.rb]: Why not just do `each` here? Why the `tap`?

File: lib/resque/scheduler/scheduling_extensions.rb [hunk 5]
```diff
@@ -121,44 +80,61 @@ def non_persistent_schedules
...
         schedule_hash.each do |name, job_spec|
           job_spec = job_spec.dup
           unless job_spec.key?('class') || job_spec.key?(:class)
             job_spec['class'] = name
           end
-          prepared_hash[name] = job_spec
+          hash[name] = job_spec
         end
-        prepared_hash
+        hash
```
Comment (Sirupsen) [inline lib/resque/scheduler/scheduling_extensions.rb]: Why rename this hash? Previous name seemed better?

File: lib/resque/scheduler/scheduling_extensions.rb (hunk 4 above)
Comment (dylanahsmith) [inline lib/resque/scheduler/scheduling_extensions.rb]: because he is modifying the returned hash

File: lib/resque/scheduler/scheduling_extensions.rb (hunk 5 above)
Comment (hkdsun) [inline lib/resque/scheduler/scheduling_extensions.rb]: yeah this slipped between iterations

**Comment by Sirupsen** [comment]
@hkdsun LGTM

Great work on keeping at it iterating on this to the simplest possible solution, which is clearly this.
